hub/config/src/main/java/uk/gov/ida/hub/config/data/ManagedEntityConfigRepository.java

```diff
@@ -36,27 +39,61 @@ public boolean has(String entityId){
-            return remoteConfigCollection.getRemoteComponent(local)
-                    .map(remote -> local.override(
-                            remote.getSignatureVerificationCertificates(),
-                            remote.getEncryptionCertificate(),
-                            CertificateOrigin.SELFSERVICE))
-                    .orElseGet(()-> {
-                        LOG.warn("Local config for '{}' expects there to be remote config but it could not be found", local.getEntityId());
-                        return local;
-                    });
-        }
-        return local;
+        return localConfigRepository.getData(entityId)
+                .map(this::overrideWithRemote);
     }
 
     public Stream<T> stream() {
         return localConfigRepository.stream();
     }
-}
\ No newline at end of file
+    
+
+    private T overrideWithRemote(T local) {
+        if (!local.isSelfService()) return local;
+        
+        return s3ConfigSource.getRemoteConfig()
+                .getRemoteComponent(local)
+                .map(remote -> getRemoteOverrides(local, remote))
+                .orElseGet(() -> logWarningAndReturnUnOverriddenConfig(local));
+    }
+
+    private T logWarningAndReturnUnOverriddenConfig(T local) {
+        LOG.warn("Local config for '{}' expects there to be remote config but it could not be found", local.getEntityId());
+        return local;
+    }
+
+    private T getRemoteOverrides(T local, RemoteComponentConfig remote) {
+        return local.override(
+                getSignatureVerificationCertificates(local, remote),
+                getEncryptionCertificate(local, remote),
+                CertificateOrigin.SELFSERVICE);
+    }
+
+    private List<String> getSignatureVerificationCertificates(T local, RemoteComponentConfig remote) {
+        try {
+            return remote.getSignatureVerificationCertificates();
+        }
+        catch (NullPointerException e){
+            LOG.warn("Remote config signing certificates missing for {}",
+                    Optional.ofNullable(local.getSignatureVerificationCertificates()
+                            .stream().map(Certificate::getSubject)
+                            .reduce("", String::join))
+                            .orElseGet(local::getEntityId));
+            throw new NoCertificateFoundException();
+        }
+    }
+
+    private String getEncryptionCertificate(T local, RemoteComponentConfig remote) {
+        try {
+            return remote.getEncryptionCertificate();
+        }
+        catch (NullPointerException e){
+            LOG.warn("Remote config encryption certificate missing for {}",
+                    local.getEncryptionCertificate().getX509Certificate()
+                            .map(X509Certificate::getSubjectDN)
+                            .map(Principal::getName)
+                            .orElseGet(local::getEntityId));
+            throw new NoCertificateFoundException();
```

Comment:
It seems weird to log a warning and then throw an error - it would be more consistent to create the `NoCertificateFoundException` with the message in the constructor instead of logging it - this way it will be correctly logged as "ERROR" and will show in the stack trace with the exception.

Reply:
I agree thanks